docs/routing.md

```diff
@@ -0,0 +1,248 @@
+
+### `error`
+
+The error phase is a rather interesting one in that it appears to need completely custom behaviour to function properly. It defines a list of source routes with status codes, usually to map a request to a custom error page.
+
+The unusual part, however, is that there is no clear way to differentiate between which should be matched, without changing the default behaviour. Instead of treating the status code property as something to be applied to the response, instead we have to treat it as something that we are trying to match with.
+
+### `hit`
+
+At the end of matching, we finally enter the `hit` phase. Here, additional headers are applied, such as the `x-matched-path` header. We enter this phase regardless of if the match ultimately resulted in a miss or not.
+
+Normally, when a match results in a miss, it will use a page at `/404` to handle the response. This would still need to have the relevant headers applied from _hitting_ a file.
+
+## Routing Process
+
+Now that we have a better understanding of the different phases, we can look at how the routing process works as a whole. The actual system that was designed to handle this is a little complex, but it can be broken down into a series of steps.
+
+Along the way, there are several times where custom behaviour has had to be introduced. This is, in part, to maximise the amount of features we support, but also because the routing system used internally by Vercel is not necessarily documented that much. Our system is custom built to serve requests, designed around our understanding of how Vercel's system should function with the Vercel build output.
+
+A simplistic overview of the routing process is as follows.
+
+- Call the handler with the request.
+- Check each phase from the build output configuration.
+  - Check each source route in a phase to see if it matches.
+  - If a match is found, apply any relevant response modifiers.
+  - Determine what to do next; return the match, or continue through routing.
+- Run the matched function, or fetch the matched static asset.
+- Return the final response.
+
+### Checking Phases
+
+The first thing that happens when a request is received, is that the router starts looking for matches in the `none` phase. It calls a function to start checking phases. This function is called recursively to check additional phases.
+
+For every single source route in a phase, it calls to check whether the route is a match, and process the relevant details (more on checking source routes can be seen below). When it receives a response from the route checker, if it encountered an error — normally from running middleware — it will break out of the routing process and return the error response. Otherwise, the only other time it disables the continuation of routing is if the route checker returns a final match, i.e. it is done for that phase.
```

Comment:
```suggestion
For every single source route in a phase, we check to see whether the route is a match, and process the relevant details (for more details on source checking see the [Checking Source Routes section](#checking-source-routes)). When we receive a response from the route checker, if we encountered an error — normally from running middleware — we break out of the routing process and return the error response. Otherwise, the only other time we disable the continuation of routing is if the route checker returns a final match.
```

Reply:
I've changed all the sentences with "it" as subject to "we", just because I don't think that specifically mentioning the function is very beneficial, but anyways if doesn't need to be "we" it could be "it" if we introduce the subject at the beginning, like "The handling process" or whatever makes more sense 🤔